### ISSUE TITLE
Can this library be used to train models from scratch?

### ISSUE THREAD

**Issue by l3utterfly** [axolotl-ai-cloud/axolotl#622]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/OpenAccess-AI-Collective/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/OpenAccess-AI-Collective/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

I would like to train smaller models from scratch (< 1B parameter). Can this library be used for it?

### ✔️ Solution

Training smaller models from scratch

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by mhenrichsen** [comment]
Yes. If the architechture of the model is supported, then it should work out of the box.

**Comment by l3utterfly** [comment]
Are there any example configs I can use to get started?

**Comment by mhenrichsen** [comment]
Find a small Llama model and use any of the provided examples.

**Comment by l3utterfly** [comment]
Sorry, can you elaborate a bit more please, I'm not sure I understand.

I would like to train a model from scratch, not fine tuning.

Why do I need to find a small model to begin? Basically, I would like to train a model from scratch by inputting the parameters, format, and raw data.

**Comment by NanoCode012** [comment]
@l3utterfly , training from scratch (aka pretraining) is just using the `completion` format but with massive amounts of data. You format the data as `{text: ".."}` (or another custom field) and use the configs in examples.

**Comment by l3utterfly** [comment]
Sorry, I'm a little new to this, so I appreciate your patience.

Say for example I want to train a Llama2 model with 1B parameters, or even smaller like 250M parameters. Or Phi model with 256M parameters.

How can I configure this in the config file? Additionally, I see in the config file it requires a "base_model" field. Should I leave those as blank when pretraining?

**Comment by mhenrichsen** [comment]
@l3utterfly you can't change the amount of parameters on the fly. If you choose a model, like Llama 1B then you can't change the model architecture and define a smaller (or larger) amount of parameters. You're stuck with the model you choose.

If you want to train a 250M parameter model, then you have to find a model that starts with that amount of parameters. Once you have found that, you replace the base_model and base_model_config with that model.

When pretraining/completion training you still need a base model to start on.

**Comment by l3utterfly** [comment]
Thanks for the answer. I have a question then, how is the first 250M model created then?

I thought I could simply define the number of layers, number of attention heads, etc. and get an "empty" model with randomized weights, then start the training from there.

**Comment by mhenrichsen** [comment]
You can, but not with Axolotl. I guess you can create a model somewhere, then upload it to hugging face and then train it in axolotl.

https://towardsdatascience.com/build-your-own-transformer-from-scratch-using-pytorch-84c850470dcb
